## API - Application Programmable Interface
* It is a software intermediary that allows two applications to talk to each other. Each time you use an app like Facebook, send an instant message, or check the weather on your phone, you’re using an API.
* It is a server that you can use to retrieve and send data to using code
![Api Img](./Api.png)

## Why API ??
* When the size of our dataset is pretty much big!
* The data is changing quickly - An example of this is stock price data. It doesn’t really make sense to regenerate a dataset and download it every minute — this will take a lot of bandwidth, and be pretty slow.

* You want a small piece of a much larger set of data. Youtube comments are one example. What if you want to just pull your own comments on Youtube? It doesn’t make much sense to download the entire Reddit database, then filter just your own comments.


## How Do I need to retrieve data From API.
* When we want to receive data from an API, we need to make a request. Requests are used all over the web.
*  when you visited this blog post, your web browser made a request to the Dataquest web server, which responded with the content of this web page. 


### Installing Requests from Python Index Packaging!
`
pip install requests
`

In [22]:
import requests
response = requests.get("https://randomuser.me/ap")
response

<Response [404]>

In [25]:
import requests
response = requests.get("https://randomuser.me/api/")
response

<Response [200]>

In [30]:
data = response.json()
data

{'results': [{'gender': 'female',
   'name': {'title': 'Miss', 'first': 'Debbie', 'last': 'Phillips'},
   'location': {'street': {'number': 723, 'name': 'Park Road'},
    'city': 'Gloucester',
    'state': 'Highlands and Islands',
    'country': 'United Kingdom',
    'postcode': 'QS19 2HE',
    'coordinates': {'latitude': '6.7380', 'longitude': '89.0713'},
    'timezone': {'offset': '+3:30', 'description': 'Tehran'}},
   'email': 'debbie.phillips@example.com',
   'login': {'uuid': '4b191958-005a-4716-b240-e6ad11c92317',
    'username': 'angrytiger492',
    'password': 'danger',
    'salt': 'avMDfASi',
    'md5': '30c0217ebceed837931511526a6c547f',
    'sha1': '6d573c97493ccb0239903fd2f08a7c964bb39e42',
    'sha256': '062352b151ff11e7bf8328feed98b873e3e5c9d74c05f960aa886ce41d776eee'},
   'dob': {'date': '1981-10-11T13:31:48.120Z', 'age': 40},
   'registered': {'date': '2017-06-06T20:28:08.301Z', 'age': 4},
   'phone': '01507 747866',
   'cell': '0797-854-613',
   'id': {'name': 'NINO', 

## API Status Codes
* Status codes are returned with every request that is made to a web server. Status codes indicate information about what happened with a request. Here are some codes that are relevant to GET requests:

* 200: Everything went okay, and the result has been returned (if any).
* 301: The server is redirecting you to a different endpoint. This can happen when a company switches domain names, or an endpoint name is changed.
* 400: The server thinks you made a bad request. This can happen when you don’t send along the right data, among other things.
* 401: The server thinks you’re not authenticated. Many APIs require login ccredentials, so this happens when you don’t send the right credentials to access an API.
* 403: The resource you’re trying to access is forbidden: you don’t have the right permissions to see it.
* 404: The resource you tried to access wasn’t found on the server.
* 503: The server is not ready to handle the request.

## Building a real-world Application using open whether map Api

In [3]:
import requests
import os
from datetime import datetime

user_api = "e31af9f4462abdba9839d05242526709"
location = input("Enter the city name: ")

complete_api_link = "https://api.openweathermap.org/data/2.5/weather?q="+location+"&appid="+user_api
api_link = requests.get(complete_api_link)
api_data = api_link.json()
api_data

Enter the city name: Trichy


{'coord': {'lon': 78.6988, 'lat': 10.8029},
 'weather': [{'id': 721,
   'main': 'Haze',
   'description': 'haze',
   'icon': '50d'}],
 'base': 'stations',
 'main': {'temp': 303.2,
  'feels_like': 305.68,
  'temp_min': 303.2,
  'temp_max': 303.2,
  'pressure': 1006,
  'humidity': 58},
 'visibility': 4000,
 'wind': {'speed': 6.17, 'deg': 280},
 'clouds': {'all': 75},
 'dt': 1626235762,
 'sys': {'type': 1,
  'id': 9227,
  'country': 'IN',
  'sunrise': 1626222587,
  'sunset': 1626268313},
 'timezone': 19800,
 'id': 7603116,
 'name': 'Trichy',
 'cod': 200}

In [4]:
#create variables to store and display data
temp_city = ((api_data['main']['temp']) - 273.15)
weather_desc = api_data['weather'][0]['description']
hmdt = api_data['main']['humidity']
wind_spd = api_data['wind']['speed']
date_time = datetime.now().strftime("%d %b %Y | %I:%M:%S %p")

print ("-------------------------------------------------------------")
print ("Weather Stats for - {}  || {}".format(location.upper(), date_time))
print ("-------------------------------------------------------------")

print ("Current temperature is: {:.2f} deg C".format(temp_city))
print ("Current weather desc  :",weather_desc)
print ("Current Humidity      :",hmdt, '%')
print ("Current wind speed    :",wind_spd ,'kmph')

-------------------------------------------------------------
Weather Stats for - TRICHY  || 14 Jul 2021 | 09:40:13 AM
-------------------------------------------------------------
Current temperature is: 30.05 deg C
Current weather desc  : haze
Current Humidity      : 58 %
Current wind speed    : 6.17 kmph
